### PR TITLE
Favortie database feature implemented

### PR DESCRIPTION
**Pull request by apoorvkhare07**
Signed-off-by: Apoorv Khare <apoorvkhare007@gmail.com>

### Description
Add Feature for selecting favourite databases similar to favourite table feature.

Before submitting pull request, please review the following checklist:

- [x] Make sure you have read our [CONTRIBUTING.md](https://github.com/phpmyadmin/phpmyadmin/blob/master/CONTRIBUTING.md) document.
- [x] Make sure you are making a pull request against the correct branch. For example, for bug fixes in a released version use the corresponding QA branch and for new features use the _master_ branch. If you have a doubt, you can ask as a comment in the bug report or on the mailing list.
- [x] Every commit has proper `Signed-off-by` line as described in our [DCO](https://github.com/phpmyadmin/phpmyadmin/blob/master/DCO). This ensures that the work you're submitting is your own creation.
- [x] Every commit has a descriptive commit message.
- [x] Every commit is needed on its own, if you have just minor fixes to previous commits, you can squash them.
- [x] Any new functionality is covered by tests.

### CODE REVIEW

**Comment by shucon** [comment]
There are failed CI tests

File: libraries/classes/Controllers/Server/DatabasesController.php [hunk 1]
```diff
@@ -247,6 +248,97 @@ public function dropDatabasesAction(array $params): array
...
+        global $cfg;
+
+        $favoriteInstance = FavoriteDatabase::getInstance('favorite_db');
+        if (isset($params['favoriteDatabases'])) {
+            $favoriteDatabases = json_decode($params['favoriteDatabases'], true);
+        } else {
+            $favoriteDatabases = [];
+        }
+        // Required to keep each user's preferences separate.
+        $user = sha1($cfg['Server']['user']);
```
Comment (williamdes) [inline libraries/classes/Controllers/Server/DatabasesController.php]: can you use `hash('sha1'` ?

File: libraries/classes/Controllers/Server/DatabasesController.php [hunk 2]
```diff
@@ -247,6 +248,97 @@ public function dropDatabasesAction(array $params): array
...
+
+        $favoriteDatabases[$user] = $favoriteInstance->getDatabases();
+
+        $json = [];
+        $json['changes'] = $changes;
+        if (! $changes) {
+            print('hello');
+            die();
+            $json['message'] = $this->template->render('components/error_message', [
+                'msg' => __("Favorite List is full!"),
```
Comment (williamdes) [inline libraries/classes/Controllers/Server/DatabasesController.php]: List does not need a capital `L` ?

File: libraries/classes/Controllers/Server/DatabasesController.php [hunk 3]
```diff
@@ -340,6 +432,7 @@ private function getDatabases(array $replicationTypes): array
 
             $databases[] = [
                 'name' => $database['SCHEMA_NAME'],
+                'name_hash' => md5($database['SCHEMA_NAME']),
```
Comment (williamdes) [inline libraries/classes/Controllers/Server/DatabasesController.php]: Can you use sha1 and hash method ?

**Comment by shucon** [comment]
You can use `https://editorconfig.org/#download` to fix trailing whitespace automatically.

File: libraries/classes/FavoriteDatabase.php [hunk 4]
```diff
@@ -0,0 +1,282 @@
...
+
+/**
+ * Handles the favorite databases.
+ *
+ * @TODO Change the release version in table pma_recent
+ * (#recent in documentation)
+ *
+ * @package PhpMyAdmin
+ */
+class FavoriteDatabase
```
Comment (williamdes) [inline libraries/classes/FavoriteDatabase.php]: That would be cool to use twig for all HTML in the class

File: libraries/classes/FavoriteDatabase.php [hunk 5]
```diff
@@ -0,0 +1,282 @@
...
+    /**
+     * Save favorite database into phpMyAdmin database.
+     *
+     * @return true|Message
+     */
+    public function saveToDb()
+    {
+        $username = $GLOBALS['cfg']['Server']['user'];
+        $sql_query
+            = " REPLACE INTO " . $this->_getPmaTable() . " (`username`, `tables`)" .
```
Comment (williamdes) [inline libraries/classes/FavoriteDatabase.php]: I think you could use `INSERT OR IGNORE INTO` ?
I think there is compatibility issues with `REPLACE INTO` ?

File: libraries/classes/FavoriteDatabase.php [hunk 6]
```diff
@@ -0,0 +1,282 @@
...
+        if ($this->_getPmaTable()) {
+            return $this->saveToDb();
+        }
+        return true;
+    }
+
+    /**
+     * Reutrn the name of the configuration storage table
+     *
+     * @return string pma table name
```
Comment (williamdes) [inline libraries/classes/FavoriteDatabase.php]: string|null

File: libraries/classes/FavoriteDatabase.php [hunk 7]
```diff
@@ -0,0 +1,282 @@
...
+     */
+    private function _getPmaTable()
```
Comment (williamdes) [inline libraries/classes/FavoriteDatabase.php]: (): ?string

File: server_databases.php [hunk 8]
```diff
@@ -30,7 +30,19 @@
 $checkUserPrivileges = new CheckUserPrivileges($dbi);
 $checkUserPrivileges->getPrivileges();
 
-if (isset($_POST['drop_selected_dbs'])
+if ( ! empty($_REQUEST['favorite_db'])) {
```
Comment (williamdes) [inline server_databases.php]: Please use `$_GET` or `$_POST`

**Review by williamdes — CHANGES_REQUESTED** [review]

File: libraries/classes/Controllers/HomeController.php [hunk 9]
```diff
@@ -82,6 +82,9 @@ public function index(): string
 
         $syncFavoriteTables = RecentFavoriteTable::getInstance('favorite')
             ->getHtmlSyncFavoriteTables();
+        
+        // $syncFavoriteTables = RecentFavoriteDb::getInstance('favorite_db')
```
Comment (shucon) [inline libraries/classes/Controllers/HomeController.php]: Remove comments

File: libraries/classes/RecentFavoriteTable.php [hunk 10]
```diff
@@ -365,6 +375,33 @@ public function getHtmlSyncFavoriteTables()
...
+     * Generate Html for sync Favorite Databases anchor. (from localStorage to pmadb)
+     *
+     * @return string
+     */
+    // public function getHtmlSyncFavoriteDatabases()
+    // {
+    //     $retval = '';
+    //     $server_id = $GLOBALS['server'];
+    //     if ($server_id == 0) {
+    //         return '';
```
Comment (shucon) [inline libraries/classes/RecentFavoriteTable.php]: Remove unnecessary comments.

File: libraries/classes/Controllers/Server/DatabasesController.php [hunk 11]
```diff
@@ -340,6 +432,7 @@ private function getDatabases(array $replicationTypes): array
 
             $databases[] = [
                 'name' => $database['SCHEMA_NAME'],
+                'name_hash' => hash(sha1($database['SCHEMA_NAME'])),
```
Comment (williamdes) [inline libraries/classes/Controllers/Server/DatabasesController.php]:
```suggestion
                'name_hash' => hash('sha1', $database['SCHEMA_NAME']),
```

**Review by williamdes — CHANGES_REQUESTED** [review]
I think you can extract the HTML to twig

**Comment by shucon** [comment]
Fix failed tests. 
Refer:
https://travis-ci.org/phpmyadmin/phpmyadmin/jobs/550498083
https://travis-ci.org/phpmyadmin/phpmyadmin/jobs/550498084

File: libraries/classes/Controllers/Server/DatabasesController.php [hunk 12]
```diff
@@ -247,6 +248,97 @@ public function dropDatabasesAction(array $params): array
...
+                $alreadyFavorite = true;  // for favorite_anchor template
+                }
+            }
+
+        $favoriteDatabases[$user] = $favoriteInstance->getDatabases();
+
+        $json = [];
+        $json['changes'] = $changes;
+        if (! $changes) {
+            print('hello');
```
Comment (shucon) [inline libraries/classes/Controllers/Server/DatabasesController.php]: This is not required.

File: libraries/classes/Controllers/Server/DatabasesController.php [hunk 13]
```diff
@@ -401,4 +495,4 @@ private function getStatisticsColumns(): array
             ],
         ];
     }
-}
+}
```
Comment (shucon) [inline libraries/classes/Controllers/Server/DatabasesController.php]:
```suggestion
}

```

File: libraries/classes/FavoriteDatabase.php [hunk 14]
```diff
@@ -0,0 +1,282 @@
...
+        $cfgRelation = $this->relation->getRelationsParam();
+        if (! empty($cfgRelation['db'])
+            && ! empty($cfgRelation[$this->_tableType])
+        ) {
+            return Util::backquote($cfgRelation['db']) . "."
+                . Util::backquote($cfgRelation[$this->_tableType]);
+        }
+        return null;
+    }
+}
```
Comment (shucon) [inline libraries/classes/FavoriteDatabase.php]:
```suggestion
}

```

File: libraries/classes/Navigation/NavigationTree.php [hunk 15]
```diff
@@ -1570,12 +1571,15 @@ private function quickWarp()
...
+        $retval .= FavoriteDatabase::getInstance('favorite_db')
+        ->getHtml();
+
         $retval .= '<div class="clearfloat"></div>';
         $retval .= '</div>';
 
         return $retval;
     }
-}
+}
```
Comment (shucon) [inline libraries/classes/Navigation/NavigationTree.php]:
```suggestion
}

```

File: server_databases.php [hunk 16]
```diff
@@ -56,4 +68,4 @@
         'sort_by' => $_REQUEST['sort_by'] ?? null,
         'sort_order' => $_REQUEST['sort_order'] ?? null,
     ]));
-}
+}
```
Comment (shucon) [inline server_databases.php]:
```suggestion
}

```

File: templates/server/databases/favorite_anchor.twig [hunk 17]
```diff
@@ -0,0 +1,7 @@
+<a id="{{ database_name }}_favorite_anchor"
+    class="ajax favorite_database_anchor"
+    href="server_databases.php{{ get_common(fav_params) }}"
+    title="{{ already_favorite ? 'Remove from Favorites'|trans : 'Add to Favorites'|trans }}"
+    data-favtargets="{{ database_name_hash }}" >
+    {{ already_favorite ? get_icon('b_favorite') : get_icon('b_no_favorite') }}
+</a>
```
Comment (shucon) [inline templates/server/databases/favorite_anchor.twig]:
```suggestion
</a>

```

File: templates/server/databases/index.twig [hunk 18]
```diff
@@ -310,4 +327,4 @@
   </div>
 {% else %}
   <p>{{ 'No databases'|trans|notice }}</p>
-{% endif %}
+{% endif %}
```
Comment (shucon) [inline templates/server/databases/index.twig]:
```suggestion
{% endif %}

```

File: libraries/classes/Controllers/HomeController.php [hunk 19]
```diff
@@ -506,4 +506,4 @@ private function checkRequirements(): void
             }
         }
     }
-}
+}
```
Comment (shucon) [inline libraries/classes/Controllers/HomeController.php]:
```suggestion
}

```

**Review by shucon — CHANGES_REQUESTED** [review]

**Comment by shucon** [comment]
@sudo-bot Deploy PR

**Comment by shucon** [comment]
Screenshot from 2019-06-27 20-37-13
It's showing an error when going to `server_databases.php`.
Also change the name of `Databases` to `Favourite Databases`.

**Comment by williamdes** [comment]
@shucon https://github.com/phpmyadmin/phpmyadmin/pull/15353#discussion_r297398225

File: libraries/classes/FavoriteDatabase.php (hunk 4 above)
Comment (apoorvkhare07) [reply libraries/classes/FavoriteDatabase.php]: @williamdes The whole side nav is currently rendered through PHP only not twig. In order to change this to twig I will have to write twig templates for the whole side nav bar.
I have written Favorite Database similar to Favorite Table already written.

**Comment by shucon** [comment]
@sudo-bot Deploy PR

**Comment by shucon** [comment]
Please fix the merge-conflicts.

**Pull request closed without merging**